text/0000-volatile-copy-and-set.md

```diff
@@ -0,0 +1,232 @@
+- Feature Name: volatile-copy-and-set
+- Start Date: 2019-04-17
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#00000](https://github.com/rust-lang/rust/issues/00000)
+
+# Summary
+[summary]: #summary
+
+Stabilize the `volatile_copy_memory`, `volatile_copy_nonoverlapping_memory`
+and `volatile_set_memory` intrinsics as `ptr::copy_volatile`,
+`ptr::copy_nonoverlapping_volatile` and `ptr::write_bytes_volatile`,
+respectively.
+
+# Motivation
+[motivation]: #motivation
+
+`ptr::read_volatile` and `ptr::write_volatile` were stabilized in RFC
+[1467](https://github.com/rust-lang/rfcs/pull/1467).  The stated motivation
+at the time was that this allowed "volatile access to memory-mapped I/O
+in stable code", something that was only previously possible using unstable
+intrinsics or "by abusing a bug in the `load` and `store` functions on
+atomic types which gives them volatile semantics
+([rust-lang/rust#30962](https://github.com/rust-lang/rust/pull/30962))."
+
+At the time, the decision was made not to also provide stable
+interfaces for the `volatile_copy_memory` or `volatile_set_memory`
+intrinsics, as they were "not used often" nor provided in C.
+However, when writing low-level code, it is sometimes also useful
+to be able to execute volatile copy and set operations.
+
+For example, when booting x86_64 "application processor" (AP) logical
+processors, code copies a sequence of instructions that for the AP to
+execute into a page in low physical memory, and then sends a startup
+inter-processor interrupt (SIPI) to the AP's local interrupt
+controller: the target interrupt vector number given in the SIPI is
+multiplied by the page size to determine the physical memory address
+where the AP should start executing.  So a SIPI sent to vector 7 of
+an AP causes that processor to begin executing instructions at
+physical memory address 0x7000.
+
+That is:
+
+```
+extern "C" {
+    fn copy_proto_page_to_phys_mem(src: usize, phys: u64);
+    fn send_init_ipi(cpu: u32);
+    fn send_sipi(cpu: u32, vector: u8);
+
+    static INIT_CODE: *const u8;
+    static INIT_CODE_LEN: usize;
+}
+
+// A contrived type for illustration; not actually useful.
+pub struct SIPIPage {
+    // Note that `bytes` is not visible outside of `SIPIPage`.
+    bytes: [u8; 4096],
+}
+
+impl SIPIPage {
+    // Note that the _only_ operation on the `bytes` field
+    // of `SIPIPage` is in `new`.  The compiler could, in
+    // theory, elide the `copy`.
```

Comment:
No it cannot, because `unsafe` code below calls `copy_proto_page_to_phys_mem` which reads from this field via a pointer to `SIPIPage`.